### PR TITLE
fix(clients): update browse iterator

### PR DESCRIPTION
**Pull request by Fluf22**
## 🧭 What and Why

🎟 JIRA Ticket: [CR-7214]

### Changes included:

- [x] java
- [x] javascript
- [x] c#
- [x] swift
- [x] ruby
- [x] php
- [x] python
- [x] kotlin (in #4059)
- [x] scala (in #4062)
- [x] go
- [ ] dart

## 🧪 Test

[CR-7214]: https://algolia.atlassian.net/browse/CR-7214?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: templates/java/api_helpers.mustache [hunk 1]
```diff
@@ -389,7 +389,7 @@ public Iterable<Rule> browseRules(String indexName, SearchRulesParams params, Re
   return AlgoliaIterableHelper.createIterable(
     () -> {
       SearchRulesResponse response = this.searchRules(indexName, params.setPage(currentPage.value), requestOptions);
-      currentPage.value = response.getNbHits() < hitsPerPage ? null : currentPage.value + 1;
```
Comment (shortcuts) [inline templates/java/api_helpers.mustache:392]: the issue is that we shouldn't check nbHits, but hits.length instead

Comment (Fluf22) [reply templates/java/api_helpers.mustache:392]: Doesn't `nbHits` reflect `hits.length`?
I think checking the page is more reliable anyway?

The only downside is it doesn't seem available for `browseSynonyms`

Comment (Fluf22) [reply templates/java/api_helpers.mustache:392]: > Doesn't nbHits reflect hits.length?

I mean: if `hitsPerPage` is 1000, and you have `x % 1000 = 0` rules, we won't stop

Comment (shortcuts) [reply templates/java/api_helpers.mustache:392]: > Doesn't nbHits reflect hits.length?

no it behaves differently based on what we fetch iirc

> I think checking the page is more reliable anyway?

idk honestly I believe it's the same, just that the legacy implem of `hits.length` have been battle tested for many years so seems safer -- but page should work too

Comment (Fluf22) [reply templates/java/api_helpers.mustache:392]: Will use the legacy implem!

File: clients/algoliasearch-client-php/lib/Iterators/RuleIterator.php
Comment (shortcuts) [inline clients/algoliasearch-client-php/lib/Iterators/RuleIterator.php:1]: does php already have the 1000 default hitsPerPage?

Comment (shortcuts) [reply clients/algoliasearch-client-php/lib/Iterators/RuleIterator.php:1]: (for browse objects)

File: templates/javascript/clients/client/api/helpers.mustache [hunk 2]
```diff
@@ -169,6 +169,10 @@ browseObjects<T>(
   }: BrowseOptions<BrowseResponse<T>> & BrowseProps,
   requestOptions?: RequestOptions
 ): Promise<BrowseResponse<T>> {
+  if (browseParams == null) {
```
Comment (shortcuts) [inline templates/javascript/clients/client/api/helpers.mustache:172]: in the spread of browseParams, could you set hitsPerPage: 1000 directly? just so if they pass a browseParams that doens't include the hitsPerPage, we still set it

**Review by shortcuts — COMMENTED** [review]
👾

File: clients/algoliasearch-client-php/lib/Iterators/RuleIterator.php
Comment (Fluf22) [reply clients/algoliasearch-client-php/lib/Iterators/RuleIterator.php:1]: Yes

Comment (shortcuts) [reply clients/algoliasearch-client-php/lib/Iterators/RuleIterator.php:1]: image

**Review by shortcuts — APPROVED** [review]
image

**Comment by Fluf22** [comment]
I'm not sure if that means:
- au quai/okay
- joli/lgtm
- beach/bitch

File: clients/algoliasearch-client-php/lib/Iterators/RuleIterator.php
Comment (Fluf22) [reply clients/algoliasearch-client-php/lib/Iterators/RuleIterator.php:1]: Took me some time to find it back 😅 But it's in the inherited class

Comment (Fluf22) [reply clients/algoliasearch-client-php/lib/Iterators/RuleIterator.php:1]: https://github.com/algolia/api-clients-automation/blob/ab5d3f630e814fce3eccd97f95f4be154648f16c/clients/algoliasearch-client-php/lib/Iterators/AbstractAlgoliaIterator.php#L50

**Comment by shortcuts** [comment]
it's NICE

**Pull request merged**